pubsublite-spark-sql-streaming/src/main/java/com/google/cloud/pubsublite/spark/SparkSourceOffset.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.pubsublite.spark;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
+import com.google.cloud.pubsublite.Partition;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public final class SparkSourceOffset
+    extends org.apache.spark.sql.sources.v2.reader.streaming.Offset {
+  private static final ObjectMapper objectMapper =
+      new ObjectMapper().configure(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS, true);
+
+  // Using a map to ensure unique partitions.
+  private final ImmutableMap<Partition, SparkPartitionOffset> partitionOffsetMap;
+
+  public SparkSourceOffset(Map<Partition, SparkPartitionOffset> map) {
+    validateMap(map);
+    this.partitionOffsetMap = ImmutableMap.copyOf(map);
+  }
+
+  private static void validateMap(Map<Partition, SparkPartitionOffset> map) {
+    map.forEach(
+        (k, v) -> {
+          assert Objects.equals(k, v.partition())
+              : "Key(Partition) and value(SparkPartitionOffset)'s partition don't match.";
+        });
+  }
+
+  public static SparkSourceOffset merge(SparkSourceOffset o1, SparkSourceOffset o2) {
+    Map<Partition, SparkPartitionOffset> result = new HashMap<>(o1.partitionOffsetMap);
+    o2.partitionOffsetMap.forEach(
+        (k, v) ->
```

Comment:
nit, run formatter? This is strange formatting

Reply:
this is what formatter gave me...